### ISSUE TITLE
Remove unfinished uploads allowing the Server to free up

### ISSUE THREAD

**Issue by sunel** [tus/tusd#176]
The Termination extension or a command to clean  unfinished uploads

**Comment by Acconut** [comment]
tusd currently has no special method for removed abandoned uploads. I would recommend you to setup a cronjob to regularly delete unused files. Here you can see the configuration we used for the master.tus.io instances: https://github.com/tus/tusd/blob/cb040c9afd8164aa18b4033deabfff1c49506586/.infra/Freyfile.hcl#L184-L190

**Comment by thePanz** [comment]
Another solution would be implement the "Expiration" extension: https://github.com/tus/tusd/issues/133

**Comment by Acconut** [comment]
Yes, that's another - although more complex - solution. However, as I said in the issue you linked: We currently have no plan to implement the Expiration extension on our own but we are always happy to assist you if you would like to contribute this feature.

**Comment by thePanz** [comment]
I'd be interested in contributing, as I have being learning GoLang recently.
The main question is architectural: how do you imagine the expiration to be handled? By an external command, or internally by the server with a cron/time.tick routine?

**Comment by Acconut** [comment]
> how do you imagine the expiration to be handled? By an external command, or internally by the server with a cron/time.tick routine?

Frankly, I am not very sure about this but that's a good question. My current tendency goes towards an internal implementation because we should not expect tusd's users to be able to build those cleanup scripts themselves as the way how tusd stores uploads is not always very accessible (which is not necessarily a bad thing. For example, it makes it easier for us to do internal refactorings if our users don't depend on certain file structures). What's your thought on this?

> the Expiration is well clear to me, as the "expiration start" should be the start of the upload (as the first byte sent), or counted from the latest byte sent?

That entirely depends on the specific implementation and its configuration. The specification intentionally does not force one of those modes as both are useful in their respective areas.

**Comment by sunel** [comment]
What i can suggest is internal implementation as a command for example

```bash

# This will be the http listener 
tusd serve 

# This can be set-up in a cron which will cleanup the files based on the given config
tusd clean --max-exp=86400

```

**Comment by kvz** [comment]
How would this work with e.g. S3 as a storage backend? If you have a million files you can't index them in a timely fashion (e.g. this cronjob would never finish before the next overlaps). 

While for S3, you could just set a retention policy to purge files after 24h. That's why I think this type of thing might be better to handle closer to the storage layer. This way tusd does not have to concern itself with the limitations and constraints of each of the different storage options out there.

**Comment by sunel** [comment]
May be this might not work for S3. Not everyone is expected to use s3.

For overlap issues a lock file might work

**Comment by kvz** [comment]
It might also not work for large NFS volumes, Google Cloud, FUSE mounted volumes, etc. So I propose to advertise people who know their storage layer handle this, because tusd can't know the intricate details about their storage layer. But offering a `tusd clean --max-exp=86400` would give people the idea that we could handle this. But then we get bug reports that we lock up people's mounts, the cleanup takes 9 weeks, or similar issues. Even if you have a lock file for this last case, a 9 weeks running cleanup for files of a day old will have unexpected consequences (like, files aren't cleaned up after 1 day).

**Comment by sunel** [comment]
If million upload is the case then custom tus server using the go source code is what i would prefer

**Comment by sunel** [comment]
@kvz you might be right

**Comment by Acconut** [comment]
For S3 (or most other cloud storage systems) we could utilize their own expirations methods to avoid holding our own index about which files should be removed.
For disk based storage we would have to use a repeating job (using cron or some own scheduling) to remove the expired files.
So I think this is not an exclusive choice but we need both approaches if we want to support this on the major data storages that we offer.

> t might also not work for large NFS volumes, Google Cloud, FUSE mounted volumes, etc. So I propose to advertise people who know their storage layer handle this, because tusd can't know the intricate details about their storage layer. But offering a tusd clean --max-exp=86400 would give people the idea that we could handle this. But then we get bug reports that we lock up people's mounts, the cleanup takes 9 weeks, or similar issues.

It's worth a consideration but I am hesitant to jump on this argument. Would you advise to not include this useful feature for 99% of tusd's users just because 1% of them use a NFS disk with millions of files? I believe that we should offer a basic support for expiration and make it clear in the documentation that it may not work optimally for extra special use cases or ultra super high volume server.

**Comment by kvz** [comment]
> Would you advise to not include this useful feature for 99% of tusd's users just because 1% of them use a NFS disk with millions of files

I think for development the 99% (or let's just call it majority) will use local/single disk volumes. But I wouldn't be surprised if in production, the majority of users actually use s3/nfs/fuse to get better scalability and durability properties. These solutions often come with trade-offs, such as that indexes aren't handled in a performant way.

So I'd suspect that we'd be building something that could potentially cause problems (outages even) in the majority of cases, not the 1%.

> For S3 (or most other cloud storage systems) we could utilize their own expirations methods to avoid holding our own index about which files should be removed.

I agree that for S3, the retention should be set. But I wonder what you mean here. Do you mean tusd should automagically to change a destination bucket's retention policy? That may require more IAM permissions than tusd is given (ideally tusd can only write/append and has no admin rights). If we/people aren't super careful, it might also apply to more files in the bucket than tusd controls, causing accidental purging of unrelated objects (?)

**Comment by Acconut** [comment]
> I think for development the 99% (or let's just call it majority) will use local/single disk volumes. But I wouldn't be surprised if in production, the majority of users actually use s3/nfs/fuse to get better scalability and durability properties. These solutions often come with trade-offs, such as that indexes aren't handled in a performant way.
So I'd suspect that we'd be building something that could potentially cause problems (outages even) in the majority of cases, not the 1%.

I agree with you that a solution which fully indexes the entire upload directory is not suitable when using NFS. However, I would keep S3 out of the argument because I believe we all agree that a separate solution for cleaning up files on S3 is needed than for cleaning up on disk (but I will talk about S3 more below).

Furthermore, I think we can still build tools for handling expiration on disk even if it may cause issues when using NFS. For me that is not a valid argument. Kitchen knife manufactures don't stop making knives just because people can cut their own fingers. Instead they make sure that you are aware of the risk. We can handle this the same way with our tool by providing proper documentation for when you should use our expiration tool or when you should roll your own solution instead. To go even further we might be able to inspect the file system type and warn the user if it's NFS/etc to account for people who do not read much documentation.

I do believe we as maintainers have a responsibility to provide solid components which do not cause harm but at the same time users should verify whether the provided components actually fit their needs and use case. And as long as we have enough documentation or help available to enable tusd's users to do this assessment, I am not much concerned about the impact. Tools can always be misused.

> Do you mean tusd should automagically to change a destination bucket's retention policy?

No, tusd should not do this. Instead, the admin should enable appropriate expiration settings for the specific S3 bucket. Then S3 will take care of cleaning up the files without causing bigger performance issues. In theory, tusd should be able to obtain the dates when a files expires and notifies the client based on that information.

**Comment by mohanklein** [comment]
@Acconut I need to further process the files (convert them to mp3) and then also inform my backend app about it. I tried post-finish hook for now to mv files to a subfolder "finished" cause I want to cleanup unresumed uploads etc. via a cronjob as you mentioned above and then create additional mp3 version of each file. But that creates the problem of the default tusd release version not serving the files anymore ... Do you get my problem? Can you give me an example of a good workflow for further processing uploaded files, such as conversion and then informing the backend app about new available files for example?

Cheers

**Comment by mohanklein** [comment]
@Acconut 
`
#!/bin/bash

echo "Upload $TUS_ID ($TUS_SIZE bytes) finished"
data=$(cat /dev/stdin)
token=$(echo $data | jq --raw-output .MetaData.token)
filename=$(echo $data | jq --raw-output .MetaData.filename)
filetype=$(echo $data | jq --raw-output .MetaData.filetype)

data_folder="../data"
finished_folder="../data/finished"
echo "Moving binary file $TUS_ID.bin ..."
mv "$data_folder/$TUS_ID.bin" "$finished_folder/$TUS_ID.bin"
mv "$data_folder/$TUS_ID.info" "$finished_folder/$TUS_ID.info"

if [[ $filetype =~ .*wav|aif.* ]]
then
    echo "Creating mp3 ..."
    lame -h -b 192 "$finished_folder/$TUS_ID.bin" "$finished_folder/mp3/$TUS_ID.mp3"
fi

echo "Done!"
`

**Comment by Acconut** [comment]
@mohanklein Please open another issue in this repository about your problems since its another topic.

In the meantime, I am not sure what you try to achieve. If you move the finished file to another location, tusd is not able to locate it anymore and is therefore not able to serve it. Could you clarify what you expect to happen?

**Comment by mohanklein** [comment]
@Acconut yeah sorry man! you're absolutely right!

**Comment by inigohu** [comment]
any update here?

**Comment by Acconut** [comment]
@inigohu No, we currently don't have many resources to work on this but your help would be appreciated if you want to give it a shot!

**Comment by Acconut** [comment]
I opened #435 to start the discussion for a possible implementation of the Expiration extension.
